### PR TITLE
CVSB-9774 - Firehose Implementation for Alert Logic

### PR DESCRIPTION
**Pull request by MartiUK**
This modifies metrics collector to be run as a firehose transformation step, allowing the logs to be dumped into an S3 bucket for Alert Logic to pull from.

### CODE REVIEW

**Review by tartino — APPROVED** [review]
LGTM

**Review by JonLazariniDD — COMMENTED** [review]
Thanks for adding some eslint rules, I think for clarity next time we can have 2 separate commits, one containing the code changes and one with the code formatting so we avoid diffing unnecessary things.
DVSA uses airbnb linting rules and comma-dangle is not configured so we do not have trailing comma, I think that is something we can turn off

File: src/cloudwatch.ts
Comment (JonLazariniDD) [inline src/cloudwatch.ts:23]: For this return I wonder whether we might want to return a type so it helps us debug the code in the future otherwise we may have to check for `instance of`/`typeof` things or make sure we don't have undefined returned. Perhaps we need a broader discussion ?

File: src/cloudwatch.ts
Comment (JonLazariniDD) [inline src/cloudwatch.ts:83]: same for this I think

File: src/dynamodb.ts
Comment (JonLazariniDD) [inline src/dynamodb.ts:27]: Does it need to be async?

File: src/handler.ts
Comment (JonLazariniDD) [inline src/handler.ts:49]: I think I will need a catch up on that one whenever you can do it please

File: tests/unit/cloudwatch.unit.test.ts
Comment (JonLazariniDD) [inline tests/unit/cloudwatch.unit.test.ts:14]: I think it is probably best if we return something, here it is undefined and could be for any reasons (function not doing its job for example), etc... I think. Happy to discuss about it.

File: src/cloudwatch.ts
Comment (MartiUK) [reply src/cloudwatch.ts:10]: BRANCH env var is expected to be set, this is to avoid a compilation error.

File: src/cloudwatch.ts
Comment (MartiUK) [reply src/cloudwatch.ts:23]: This used to return a string which would then only be logged, since the logging is done in this method anyway there is no point to return a value.

File: src/handler.ts
Comment (MartiUK) [reply src/handler.ts:49]: The code was written before would get batches of logs for a single microservice, now using firehose, there will likely be multiple microservice logs interspersed. There is no benefit to sending the same metrics milliseconds after the last.

File: src/dynamodb.ts
Comment (MartiUK) [reply src/dynamodb.ts:27]: No

File: tests/unit/cloudwatch.unit.test.ts
Comment (MartiUK) [reply tests/unit/cloudwatch.unit.test.ts:14]: See above explanations on these.

Comment (MartiUK) [reply tests/unit/cloudwatch.unit.test.ts:14]: This is specifically to expect a void promise.

File: src/dynamodb.ts
Comment (JonLazariniDD) [reply src/dynamodb.ts:27]: can we make it sync if we don't need async operations for it please?

**Review by JonLazariniDD — COMMENTED** [review]
one for later but this is the DVSA standard in terms of linting

https://github.com/dvsa/eslint-config-ts#rules

File: src/cloudwatch.ts
Comment (JonLazariniDD) [reply src/cloudwatch.ts:10]: I leave it to you but I thought it would easier for the dev if "local" was set as a default branch in case no BRANCH is passed from node (can't see a script in the package.json to run the service which would help any maintainer)

File: src/cloudwatch.ts
Comment (JonLazariniDD) [reply src/cloudwatch.ts:23]: I understand we are consuming it for now but i think it is better if we can return the timeoutCount from the function as it seems it what changes depending on the logEvents (and this is what we publish to amazon cloudwatch. This will make the unit testing better as we are likely to mock that and if for whatever reasons this was not mocked properly we would assert on undefined and will make it hard to to know whether the unit test passes because the function did its job or because the mock is undefined (or any other weird undefined behaviour happening with js)

File: src/cloudwatch.ts
Comment (MartiUK) [reply src/cloudwatch.ts:12]: This is defaulting to local, if BRANCH env var is null it will be "local"

**Review by JonLazariniDD — APPROVED** [review]

**Review by JonLazariniDD — APPROVED** [review]

**Review by tartino — DISMISSED** [review]
LGTM

**Review by JonLazariniDD — APPROVED** [review]

**Review by tartino — APPROVED** [review]

**Pull request merged**
